internal/internal.go

```diff
@@ -695,36 +725,131 @@ func getParsedPathAndQuery(path string) ([]interface{}, map[string]interface{},
 	return parsedPathInterface, parsedQueryInterface, nil
 }
 
-func getParsedBody(headers map[string]string, body string) (interface{}, bool, error) {
-	if body == "" {
-		return nil, false, nil
-	}
-
-	if val, ok := headers["content-length"]; ok {
-		cl, err := strconv.ParseInt(val, 10, 64)
-		if err != nil {
-			return nil, false, err
-		}
-
-		if cl != -1 && cl > int64(len(body)) {
-			return nil, true, nil
-		}
-	}
-
+func getParsedBody(headers map[string]string, body string, rawBody []byte, parsedPath []interface{}, protoSet *protoregistry.Files) (interface{}, bool, error) {
 	var data interface{}
 
 	if val, ok := headers["content-type"]; ok {
 		if strings.Contains(val, "application/json") {
+
+			if body == "" {
+				return nil, false, nil
+			}
+
+			if val, ok := headers["content-length"]; ok {
+				cl, err := strconv.ParseInt(val, 10, 64)
+				if err != nil {
+					return nil, false, err
+				}
+				if cl != -1 && cl > int64(len(body)) {
+					return nil, true, nil
+				}
+			}
+
 			err := util.Unmarshal([]byte(body), &data)
 			if err != nil {
 				return nil, false, err
 			}
+		} else if strings.Contains(val, "application/grpc") {
+
+			if protoSet == nil {
+				return nil, false, nil
+			}
+
+			if len(rawBody) == 0 {
+				return nil, false, fmt.Errorf("invalid raw body")
```

Comment:
🚧 This has to change: With the other corner cases, we've left it to the policy author to decide what to do -- if the body couldn't be parsed, or if it was truncated; **this case here** would be hit if opa-envoy-plugin was configured to parse the protobuf bodies but didn't get one. It should behave the same, `return nil, false, nil`.